packages/block-library/src/list/transforms.js

```diff
@@ -72,102 +77,43 @@ const transforms = {
 				ol: getListContentSchema( args ).ol,
 				ul: getListContentSchema( args ).ul,
 			} ),
-			transform( node ) {
-				const attributes = {
-					ordered: node.nodeName === 'OL',
-					anchor: node.id === '' ? undefined : node.id,
-				};
-
-				if ( attributes.ordered ) {
-					const type = node.getAttribute( 'type' );
-
-					if ( type ) {
-						attributes.type = type;
-					}
-
-					if ( node.getAttribute( 'reversed' ) !== null ) {
-						attributes.reversed = true;
-					}
-
-					const start = parseInt( node.getAttribute( 'start' ), 10 );
-
-					if (
-						! isNaN( start ) &&
-						// start=1 only makes sense if the list is reversed.
-						( start !== 1 || attributes.reversed )
-					) {
-						attributes.start = start;
-					}
-				}
-
-				return createBlock( 'core/list', {
-					...getBlockAttributes( 'core/list', node.outerHTML ),
-					...attributes,
-				} );
-			},
+			transform: createListBlockFromDOMElement,
 		},
 		...[ '*', '-' ].map( ( prefix ) => ( {
 			type: 'prefix',
 			prefix,
 			transform( content ) {
-				return createBlock( 'core/list', {
-					values: `<li>${ content }</li>`,
-				} );
+				return createBlock( 'core/list', {}, [
+					createBlock( 'core/list-item', { content } ),
+				] );
 			},
 		} ) ),
 		...[ '1.', '1)' ].map( ( prefix ) => ( {
 			type: 'prefix',
 			prefix,
 			transform( content ) {
-				return createBlock( 'core/list', {
-					ordered: true,
-					values: `<li>${ content }</li>`,
-				} );
+				return createBlock(
+					'core/list',
+					{
+						ordered: true,
+					},
+					[ createBlock( 'core/list-item', { content } ) ]
+				);
 			},
 		} ) ),
 	],
 	to: [
-		{
+		...[ 'core/paragraph', 'core/heading' ].map( ( block ) => ( {
 			type: 'block',
-			blocks: [ 'core/paragraph' ],
-			transform: ( { values } ) =>
-				split(
-					create( {
-						html: values,
-						multilineTag: 'li',
-						multilineWrapperTags: [ 'ul', 'ol' ],
-					} ),
-					__UNSTABLE_LINE_SEPARATOR
-				).map( ( piece ) =>
-					createBlock( 'core/paragraph', {
-						content: toHTMLString( { value: piece } ),
+			blocks: [ block ],
+			transform: ( _attributes, childBlocks ) => {
+				return getListContentFlat( childBlocks ).map( ( content ) =>
+					createBlock( block, {
+						content,
 					} )
-				),
-		},
-		{
-			type: 'block',
-			blocks: [ 'core/heading' ],
-			transform: ( { values } ) =>
-				split(
-					create( {
-						html: values,
-						multilineTag: 'li',
-						multilineWrapperTags: [ 'ul', 'ol' ],
-					} ),
-					__UNSTABLE_LINE_SEPARATOR
-				).map( ( piece ) =>
-					createBlock( 'core/heading', {
-						content: toHTMLString( { value: piece } ),
-					} )
-				),
-		},
-		{
```

Comment:
This transform has been removed. At first sight, it looks like it may be not useful (it creates a TOC without considering the list contents), though upon testing it seems to support a specific use case: when an user creates a manually curated list to maintain a TOC. By transforming that list to a TOC block it'll be automatically generated based on the headings of the post.

Reply:
Yes, I removed it because it didn't make sense to me. The list content will be removed unexpectedly, which goes against us always trying to preserve any content with transforms. I don't understand the specific use case. You can easily create a new ToC block that does the same thing and explicitly delete the list if it's not useful anymore.

Reply:
@ellatrix Yeah, I only added that transform for the sake of convenience, and because I had heard somewhere that all transforms should ideally be reversible. https://github.com/WordPress/gutenberg/pull/29739#issuecomment-1083394967

I agree that it doesn't really make sense.

Reply:
Your rationale makes a lot of sense. It wasn't clear to me why it was there in the first place. I guess my concern is that it's a behavioral change that is unrelated to this refactoring to inner blocks. I'm not sure how people would perceive it. I'd perhaps address this separately and ask feedback, though I don't feel strongly about it. I trust whatever you feel is the best direction.

Reply:
I guess if the original author is on board we're good to go :)